File: samples/dotnetcore/GroupsAndTags/GroupsAndTags/Groups.cs
// ===============================================================================
// Alachisoft (R) NCache Sample Code.
// ===============================================================================
// Copyright © Alachisoft.  All rights reserved.
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
// OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
// LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE.
// ===============================================================================

using System;
using System.Collections;
using Alachisoft.NCache.Sample.Data;
using System.Configuration;
using Alachisoft.NCache.Client;
using System.Collections.Generic;

namespace Alachisoft.NCache.Samples
{
    /// <summary>
    /// Class that demonstrates the usage of groups in NCache
    /// </summary>
    public class Groups
    {
        private static ICache _cache;

        public static void Run()
        {
            // Initialize the cache
            InitializeCache();

            //Adding item in same group        
            AddItems();

            // Getting group data
            // Will return nine items
            GetItemsByGroup();

            // getGroupKeys is yet another function to retrive group data.
            // It however requires multiple iterations to retrive actual data
            // 1) To get List of Keys and 2) TO get items for the return List of Keys

            // Updating items in groups
            // Item is updated at the specified group
            UpdateItem();

            // Remove group data
            RemoveGroupData();

            // Dispose cache once done
            _cache.Dispose();
        }

        /// <summary>
        /// This method initializes the cache.
        /// </summary>
        private static void InitializeCache()
        {
            string cache = ConfigurationManager.AppSettings["CacheId"];

            if (String.IsNullOrEmpty(cache))
            {
                Console.WriteLine("The Cache Name cannot be null or empty.");
                return;
            }

            // Initialize an instance of the cache to begin performing operations:
            _cache = NCache.Client.CacheManager.GetCache(cache);
            _cache.Clear();
            Console.WriteLine("Cache initialized successfully");
        }

        /// <summary>
        /// This method add items in the cache.
        /// </summary>
        private static void AddItems()
        {
            // Mobile Items

            AddGroupedDatatoCache(1, "PhoneSamsung", "Mobiles");
            AddGroupedDatatoCache(2, "HTCPhone", "Mobiles");
            AddGroupedDatatoCache(3, "NokiaPhone", "Mobiles");

            // Laptop Items

            AddGroupedDatatoCache(4, "AcerLaptop", "Laptops");
            AddGroupedDatatoCache(5, "HPLaptop", "Laptops");
            AddGroupedDatatoCache(6, "DellLaptop", "Laptops");

            // small electronics items

            AddGroupedDatatoCache(7, "HairDryer", "SmallElectronics");
            AddGroupedDatatoCache(8, "VaccumCleaner", "SmallElectronics");
            AddGroupedDatatoCache(9, "Iron", "SmallElectronics");



            Console.WriteLine("Items added in cache.");
        }

        private static void AddGroupedDatatoCache(int productId, string productName, string group, string className = "Elec", string category = "Elec")
        {
            Product product = new Product() { Id = productId, Name = productName, ClassName = className, Category = category };

            // create object key
            string key = "product:" + product.Id.ToString();

            // create CacheItem with your desired object
            CacheItem item = new CacheItem(product);

            // assign group to CacheItem object
            item.Group = group;

            // add CacheItem object to cache
            _cache.Add(key, item);
        }



        /// <summary>
        /// This method fetches items from the cache using groups
        /// </summary>
        private static void GetItemsByGroup()
        {
            IDictionary<string, Product> items = _cache.SearchService.GetGroupData<Product>("Mobiles");
            if (items.Count > 0)
            {
                Console.WriteLine("Item count: " + items.Count);
                Console.WriteLine("Following Products are found in group 'Mobiles'");
                IEnumerator itor = items.Values.GetEnumerator();
                while (itor.MoveNext())
                {
                    Console.WriteLine(((Product)itor.Current).Name);
                }
                Console.WriteLine();
            }
        }

        /// <summary>
        /// This mehtod updates items in the cache.
        /// </summary>
        private static void UpdateItem()
        {
            Product product = new Product() { Id = 8, Name = "PanaSonicIron" };

            CacheItem panasonicIron = new CacheItem(product);
            panasonicIron.Group = "SmallElectronics";

            string key = "product:" + product.Id.ToString();
            _cache.Insert(key, panasonicIron);
        }

        /// <summary>
        /// This method removes items from the cache that belong to a specific group
        /// </summary>
        private static void RemoveGroupData()
        {
            Console.WriteLine("Item count: " + _cache.Count); // Itemcount = 6
            _cache.SearchService.RemoveGroupData("Laptops"); // Will remove all items from cache based on group Laptops

            Console.WriteLine("Item count: " + _cache.Count); // Itemcount = 0
        }
    }
}

File: nodejs/groupsandtags/src/groups.js
const {CacheManager,CacheItem} = require('ncache-client');
let cache = null;
const config = require('./app.config.json');
async function Run() {
    // Initialize cache 
    await InitializeCache();
    //Adding item in same group        
    await AddItems();

    // Getting group data
    // Will return nine items
    await GetItemsByGroup();

    // getGroupKeys is yet another function to retrive group data.
    // It however requires multiple iterations to retrive actual data
    // 1) To get List of Keys and 2) TO get items for the return List of Keys

    // Updating items in groups
    // Item is updated at the specified group
    await UpdateItem();

    // Remove group data
    await RemoveGroupData();

    //close cache;
    await cache.close();

}

async function InitializeCache() {
    // Get Cache Name
    const cacheName = config.cacheProperties.cacheName
    if (!cacheName) {
        console.log("The CacheID cannot be null or empty.");
        return;
    }
    // Initialize an instance of the cache to begin performing operations:
    cache = await CacheManager.getCache(cacheName);

    //clear cache data
    await cache.clear();

    // Print output on console
    console.log(`Cache ${cacheName} is initialized.`);
            
}

async function AddItems() {
    await Promise.all([
        //Mobile items
        AddGroupedDatatoCache(1, "PhoneSamsung", "Mobiles"),
        AddGroupedDatatoCache(2, "HTCPhone", "Mobiles"),
        AddGroupedDatatoCache(3, "NokiaPhone", "Mobiles"),

            // Laptop Items

        AddGroupedDatatoCache(4, "AcerLaptop", "Laptops"),
        AddGroupedDatatoCache(5, "HPLaptop", "Laptops"),
        AddGroupedDatatoCache(6, "DellLaptop", "Laptops"),

        // small electronics items

        AddGroupedDatatoCache(7, "HairDryer", "SmallElectronics"),
        AddGroupedDatatoCache(8, "VaccumCleaner", "SmallElectronics"),
        AddGroupedDatatoCache(9, "Iron", "SmallElectronics")
    ]);

    console.log("items added in cache");
}
async function AddGroupedDatatoCache(productId, productName, group, className = "Elec", category = "Elec")
{
    const product = { Id : productId, Name : productName, ClassName : className, Category : category };

    // create object key
    const key = "product:" + product.Id;

    // create CacheItem with your desired object
    const  item = new CacheItem(product,"Product");

    // assign group to CacheItem object
    item.setGroup(group);

    // add CacheItem object to cache
    await cache.add(key, item);
}
async function GetItemsByGroup() {
    const searchService = await cache.getSearchService();
    const items = await searchService.getGroupData("Mobiles");
    if (items.size > 0) {
        console.log("Item count: " + items.size);
        console.log("Following Products are found in group 'Mobiles'");
        for (const item of items.values()) {
            console.log(item.toStringSync());
        }
    }
}

async function UpdateItem() {
    const product = { Id : 8, Name : "PanaSonicIron"};
    const panasonicIron = new CacheItem(product,"Product");
    panasonicIron.setGroup("SmallElectronics");
    const key = "product:" + product.Id;
    await cache.insert(key, panasonicIron);
}

async function RemoveGroupData() {
    let count = await cache.getCount()
    console.log("Item count: " + count); 

    const searchService = await cache.getSearchService();
    await searchService.removeGroupData("Laptops");

    count = await cache.getCount()
    console.log("Item count: " + count); 
}
module.exports = {
    Run
}